### PR TITLE
Improve status block of UnifiedPushServer CRD

### PR DESCRIPTION
**Pull request by grdryn**
JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

## Progress

- [x] Updated status in CRD
- [x] Added manageError function to set status & fire event
- [x] Add secondary resources to map in CR status
- [x] Set "ready" status appropriately

## Checklist:

- [x] Code has been tested locally by PR requester
- [x] Changes have been successfully verified by another team member

## Verification steps

I've got this all already deployed on an OSD 4 dev cluster, so if you'd like to check it out there, let me know, it might save some time :smiley: 

### Upgrade
- [ ] Deploy operator & UPS from master:
```
git checkout master
make install
# wait for everything to become ready
```
- [ ] Upgrade to configuration from this PR
```
git checkout AEROGEAR-9958
sed -i -e "s|quay.io/aerogear/unifiedpush-operator:.*|quay.io/aerogear/unifiedpush-operator:PR-84|g" deploy/operator.yaml
make install
# ^re-applies all CRDs so they've got the new stuff
```

- [ ] Make sure there are no persistent/blocking errors in the operator logs

### Fresh install
- [ ] Delete the UPS instance:
```
kubectl delete ups --all
```

- [ ] Create a new UPS
```
kubectl apply -f deploy/crds/push_v1alpha1_unifiedpushserver_cr.yaml
```

- [ ] Make sure everything comes up as expected, and there aren't any persistent/blocking errors in the operator logs

### CODE REVIEW

File: pkg/controller/unifiedpushserver/unifiedpushserver_controller.go [hunk 1]
```diff
@@ -7,13 +7,14 @@ import (
...
-	"k8s.io/client-go/rest"
-
 	"github.com/aerogear/unifiedpush-operator/pkg/constants"
 
 	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
 
+	"github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
```
Comment (grdryn) [inline pkg/controller/unifiedpushserver/unifiedpushserver_controller.go]: hmm, I'll fix this dupe

File: pkg/controller/unifiedpushserver/unifiedpushserver_controller.go [hunk 2]
```diff
@@ -987,7 +1028,7 @@ func (r *ReconcileUnifiedPushServer) Reconcile(request reconcile.Request) (recon
 		err := controllerutil.SetControllerReference(instance, serviceMonitor, r.scheme)
 		return err
 	})
-	if err != nil {
+	if _, isNoKindMatchError := err.(*meta.NoKindMatchError); err != nil && !isNoKindMatchError {
 		return reconcile.Result{}, err
```
Comment (davidffrench) [inline pkg/controller/unifiedpushserver/unifiedpushserver_controller.go]: Is there a reason this and the ones below are not using `r. manageError`?

**Review by davidffrench — COMMENTED** [review]
This looks really good @grdryn , great work! Just one very small comment

File: pkg/controller/unifiedpushserver/unifiedpushserver_controller.go (hunk 2 above)
Comment (grdryn) [reply pkg/controller/unifiedpushserver/unifiedpushserver_controller.go]: Good spot, thanks -- this is new code that showed up after a rebase & I didn't update it (apart from the fix to ignore if it's a NoKindMatchError). I'll update it :+1:

File: deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml [hunk 3]
```diff
@@ -85,10 +85,31 @@ spec:
...
+                  type: string
+                type: array
+              description: 'SecondaryResources is a map of all the secondary resources
+                types and names created for this CR.  e.g "Deployment": [ "DeploymentName1",
+                "DeploymentName2" ]'
+              type: object
           required:
           - phase
+          - message
+          - ready
```
Comment (grdryn) [inline deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml]: @davidffrench I'm thinking that maybe these shouldn't be "required" for this version of the CRD, wdyt? The case I'm thinking of is where someone applies the CRDs but can't actually upgrade the operator immediately for one reason or another (or have to roll back) -- the existing version of the operator won't be able to reconcile the updated CRD if we make these required.

**Comment by secondsun** [comment]
@grdryn Does this require crc or can it work on minishift?

**Comment by grdryn** [comment]
@secondsun I can't think of any reason why it wouldn't work on minishift.

File: deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml (hunk 3 above)
Comment (davidffrench) [reply deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml]: I would love to know what happens in this scenario for 'required' status fields. The scenario you are outlining makes sense to not make them mandatory.

**Comment by davidffrench** [comment]
@secondsun Seconding what @grdryn said, no reason it shouldn't work.

The only reason I have come across for it to not work on k8s 1.11 is a structural schema used in the CRD but that doesn't look to be the case here.

File: deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml (hunk 3 above)
Comment (grdryn) [reply deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml]: I tested it out, and got the following error in one case:

<details>
{"level":"error","ts":1574341351.4474332,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"unifiedpushserver-controller","request":"mobile-unifiedpush/unifiedpush","error":"UnifiedPushServer.push.aerogear.org \"unifiedpush\" is invalid: []: Invalid value: map[string]interface {}{\"kind\":\"UnifiedPushServer\", \"metadata\":map[string]interface {}{\"clusterName\":\"\", \"annotations\":map[string]interface {}{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"push.aerogear.org/v1alpha1\\\",\\\"kind\\\":\\\"UnifiedPushServer\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"generation\\\":1,\\\"name\\\":\\\"unifiedpush\\\",\\\"namespace\\\":\\\"mobile-unifiedpush\\\",\\\"resourceVersion\\\":\\\"\\\",\\\"selfLink\\\":\\\"/apis/push.aerogear.org/v1alpha1/namespaces/mobile-unifiedpush/unifiedpushservers/unifiedpush\\\",\\\"uid\\\":\\\"\\\"},\\\"spec\\\":{\\\"useMessageBroker\\\":false}}\\n\"}, \"resourceVersion\":\"245971\", \"name\":\"unifiedpush\", \"namespace\":\"mobile-unifiedpush\", \"uid\":\"2306dd4a-0c5f-11ea-9c2f-16c554726cf3\", \"generation\":1, \"creationTimestamp\":\"2019-11-21T13:02:13Z\"}, \"status\":map[string]interface {}{\"phase\":\"Provisioning\"}, \"spec\":map[string]interface {}{\"useMessageBroker\":false}, \"apiVersion\":\"push.aerogear.org/v1alpha1\"}: validation failure list:\nstatus.message in body is required\nstatus.ready in body is required","stacktrace":"github.com/aerogear/unifiedpush-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
</details>

^ which basically boils down to `validation failure list:\nstatus.message in body is required\nstatus.ready in body is required` if you care to look closely enough.

I'll update these to make them optional :)

**Comment by secondsun** [comment]
Code looks good, code runs good, here's the ups cr after all the updates are applied.

```
[summerspittman@localhost unifiedpush-operator]$ oc describe ups
Name:         example-unifiedpushserver
Namespace:    unifiedpush
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"push.aerogear.org/v1alpha1","kind":"UnifiedPushServer","metadata":{"annotations":{},"name":"example-unifiedpushserver","namespace":"unif...
API Version:  push.aerogear.org/v1alpha1
Kind:         UnifiedPushServer
Metadata:
  Creation Timestamp:  2019-11-21T13:57:43Z
  Generation:          1
  Resource Version:    5083
  Self Link:           /apis/push.aerogear.org/v1alpha1/namespaces/unifiedpush/unifiedpushservers/example-unifiedpushserver
  UID:                 e3b331a8-0c66-11ea-afb5-52540025c695
Spec:
  Use Message Broker:  false
Status:
  Message:  
  Phase:    Reconciling
  Ready:    true
  Secondary Resources:
    Deployment:
      example-unifiedpushserver-postgresql
      example-unifiedpushserver
    Persistent Volume Claim:
      example-unifiedpushserver-postgresql
    Route:
      example-unifiedpushserver-unifiedpush-proxy
    Secret:
      example-unifiedpushserver-postgresql
    Service:
      example-unifiedpushserver-postgresql
      example-unifiedpushserver-unifiedpush-proxy
      example-unifiedpushserver-unifiedpush
    Service Account:
      example-unifiedpushserver
Events:  <none>

```

**Review by secondsun — APPROVED** [review]

**Comment by grdryn** [comment]
Thanks a lot for the review & verification @secondsun! As mentioned in another comment, I'm going to make a tiny change to just make the new status fields optional, so it's not breaking backwards compatibility with the older version of the operator (I actually thought that I'd already pushed that before lunch, but clearly not). I don't consider that as needing a re-verification though.

**Review by davidffrench — APPROVED** [review]
New changes LGTM. Great work @grdryn

**Pull request merged**